tests/tests/integration_tests.rs

```diff
@@ -274,14 +274,13 @@ async fn ws_close_pending_subscription_when_server_terminated() {
 
 	assert!(matches!(sub.next().await, Ok(Some(_))));
 
-	handle.stop().await.unwrap();
-	handle.wait_for_stop().await;
+	handle.stop().unwrap().await;
 
 	let sub2: Result<Subscription<String>, _> =
 		c1.subscribe("subscribe_hello", JsonRpcParams::NoParams, "unsubscribe_hello").await;
 
 	// no new request should be accepted.
 	assert!(matches!(sub2, Err(_)));
 	// the already established subscription should also be closed.
-	assert!(matches!(sub.next().await, Ok(None)));
+	// assert!(matches!(sub.next().await, Ok(None)));
```

Comment:
I see, it worked as the background task created a channel to send & receive data, then when the receiver was dropped [here](https://github.com/paritytech/jsonrpsee/blob/master/ws-server/src/server.rs#L245-#L262) caused the `SubscriptionSink::send` to fail and send a `SubscriptionClosed message` to the server.

However, it seems like the `Server` terminates the connection before that happens on `master` (not what I remember).

In your case it could simply be timing thingy or some regression with the changed `StopHandle`. I can test it on your branch later.

Reply:
I just debugged this and on this branch it actually sends on two responses on the subscription before the connection is terminated. Such that the `assertion on line 285 will be Some(_)` but after polling it one more time it will be return `Ok(None)`. Thus, timing issue/badly written test ^^

```bash
[2021-08-12T15:12:18Z DEBUG jsonrpsee_ws_client::transport] send: {"jsonrpc":"2.0","method":"subscribe_hello","id":0,"params":null}
[2021-08-12T15:12:18Z DEBUG jsonrpsee_ws_server::server] recv: JsonRpcRequest { jsonrpc: TwoPointZero, id: Number(0), method: "subscribe_hello", params: None }
[2021-08-12T15:12:18Z DEBUG jsonrpsee_ws_server::server] send: {"jsonrpc":"2.0","result":3287776652631530,"id":0}
[2021-08-12T15:12:18Z DEBUG jsonrpsee_ws_server::server] send: {"jsonrpc":"2.0","method":"subscribe_hello","params":{"subscription":3287776652631530,"result":"hello from subscription"}}
[2021-08-12T15:12:18Z DEBUG jsonrpsee_ws_client::client] [backend]: recv method_call JsonRpcResponse { jsonrpc: TwoPointZero, result: Number(3287776652631530), id: Number(0) }
[2021-08-12T15:12:18Z DEBUG jsonrpsee_ws_client::client] [backend]: recv subscription JsonRpcNotification { jsonrpc: TwoPointZero, method: "subscribe_hello", params: JsonRpcSubscriptionParams { subscription: Num(3287776652631530), result: String("hello from subscription") } }
[2021-08-12T15:12:18Z DEBUG jsonrpsee_ws_server::server] send: {"jsonrpc":"2.0","method":"subscribe_hello","params":{"subscription":3287776652631530,"result":"hello from subscription"}}
[2021-08-12T15:12:18Z DEBUG jsonrpsee_ws_client::client] [backend]: recv subscription JsonRpcNotification { jsonrpc: TwoPointZero, method: "subscribe_hello", params: JsonRpcSubscriptionParams { subscription: Num(3287776652631530), result: String("hello from subscription") } }
[2021-08-12T15:12:18Z ERROR jsonrpsee_ws_client::client] Error: Connection(Closed) terminating client
```

on master we get this:

```bash
[2021-08-12T15:17:24Z DEBUG jsonrpsee_ws_client::transport] Connecting to target: Target { sockaddrs: [127.0.0.1:39995], host: "127.0.0.1", host_header: "127.0.0.1:39995", mode: Plain, path_and_query: "/" }
[2021-08-12T15:17:24Z DEBUG jsonrpsee_ws_client::transport] send: {"jsonrpc":"2.0","method":"subscribe_hello","id":0,"params":null}
[2021-08-12T15:17:24Z DEBUG jsonrpsee_ws_server::server] recv: JsonRpcRequest { jsonrpc: TwoPointZero, id: Number(0), method: "subscribe_hello", params: None }
[2021-08-12T15:17:24Z DEBUG jsonrpsee_ws_server::server] send: {"jsonrpc":"2.0","result":2896007276381328,"id":0}
[2021-08-12T15:17:24Z DEBUG jsonrpsee_ws_server::server] send: {"jsonrpc":"2.0","method":"subscribe_hello","params":{"subscription":2896007276381328,"result":"hello from subscription"}}
[2021-08-12T15:17:24Z DEBUG jsonrpsee_ws_client::client] [backend]: recv method_call JsonRpcResponse { jsonrpc: TwoPointZero, result: Number(2896007276381328), id: Number(0) }
[2021-08-12T15:17:24Z DEBUG jsonrpsee_ws_client::client] [backend]: recv subscription JsonRpcNotification { jsonrpc: TwoPointZero, method: "subscribe_hello", params: JsonRpcSubscriptionParams { subscription: Num(2896007276381328), result: String("hello from subscription") } }
[2021-08-12T15:17:24Z DEBUG jsonrpsee_ws_client::transport] send: {"jsonrpc":"2.0","method":"subscribe_hello","id":2,"params":null}
[2021-08-12T15:17:24Z ERROR jsonrpsee_ws_client::client] Error: Connection(Closed) terminating client
test ws_close_pending_subscription_when_server_terminated ... ok
```

Not really sure why but maybe you have an idea?! Maybe the closing stuff is slower?

Reply:
Ok, I've updated the test to allow for one message before the server shuts down. Kind of makes sense now that everything is 100% gracefully shutting down.